tree/tree/inc/TLeaf.h

```diff
@@ -49,57 +49,68 @@ class TLeaf : public TNamed {
-   virtual void     FillBasket(TBuffer& b);
-   TBranch         *GetBranch() const { return fBranch; }
-   virtual TLeaf   *GetLeafCount() const { return fLeafCount; }
-   virtual TLeaf   *GetLeafCounter(Int_t& countval) const;
-   virtual Int_t    GetLen() const;
-   virtual Int_t    GetLenStatic() const { return fLen; }
-   virtual Int_t    GetLenType() const { return fLenType; }
-   virtual Int_t    GetMaximum() const { return 0; }
-   virtual Int_t    GetMinimum() const { return 0; }
-   virtual Int_t    GetNdata() const { return fNdata; }
-   virtual Int_t    GetOffset() const { return fOffset; }
-   virtual void    *GetValuePointer() const { return 0; }
-   virtual const char *GetTypeName() const { return ""; }
-
-   virtual Double_t GetValue(Int_t i = 0) const;
-   virtual Long64_t GetValueLong64(Int_t i = 0) const { return GetValue(i); } //overload only when it matters.
-   virtual LongDouble_t GetValueLongDouble(Int_t i = 0) const { return GetValue(i); } // overload only when it matters.
-   template <typename T > T GetTypedValue(Int_t i = 0) const { return GetValueHelper<T>::Exec(this, i); }
-
-   virtual void     Import(TClonesArray*, Int_t) {}
-   virtual Bool_t   IsOnTerminalBranch() const { return kTRUE; }
-   virtual Bool_t   IsRange() const { return fIsRange; }
-   virtual Bool_t   IsUnsigned() const { return fIsUnsigned; }
-   virtual void     PrintValue(Int_t i = 0) const;
-   virtual void     ReadBasket(TBuffer&) {}
-   virtual void     ReadBasketExport(TBuffer&, TClonesArray*, Int_t) {}
-   virtual void     ReadValue(std::istream& /*s*/, Char_t /*delim*/ = ' ') {
-      Error("ReadValue", "Not implemented!");
-   }
-           Int_t    ResetAddress(void* add, Bool_t destructor = kFALSE);
-   virtual void     SetAddress(void* add = 0);
-   virtual void     SetBranch(TBranch* branch) { fBranch = branch; }
-   virtual void     SetLeafCount(TLeaf* leaf);
-   virtual void     SetLen(Int_t len = 1) { fLen = len; }
-   virtual void     SetOffset(Int_t offset = 0) { fOffset = offset; }
-   virtual void     SetRange(Bool_t range = kTRUE) { fIsRange = range; }
-   virtual void     SetUnsigned() { fIsUnsigned = kTRUE; }
-
-   ClassDef(TLeaf,2);  //Leaf: description of a Branch data type
+  Int_t *GenerateOffsetArrayBase(Int_t base, Int_t events) const; // For leaves containing fixed-size objects (no
+                                                                  // polymorphism!), this will generate an appropriate
+                                                                  // offset array.
+
+  public:
+  enum EStatusBits {
+     kIndirectAddress = BIT(11), ///< Data member is a pointer to an array of basic types.
+     kNewValue = BIT(12)         ///< Set if we own the value buffer and so must delete it ourselves.
+  };
+
+  TLeaf();
+  TLeaf(TBranch *parent, const char *name, const char *type);
+  virtual ~TLeaf();
+
+  virtual void Browse(TBrowser *b);
+  virtual Bool_t CanGenerateOffsetArray()
+  {
```

Comment:
In addition the underlying content must be of 'fixed' size (i.e. not a non-split object itself contains pointers or collections).

Reply:
That is true.  Notice how I set:
```
virtual Bool_t CanGenerateOffsetArray() { return false; }
```
for `TLeafO`.

In a follow-up commit, we'll need to delve into the various use cases inside `TLeafO` where skipping offset array can be done for non-primitive objects.

Reply:
Humm .. you will also eventually have to deal with TLeafElement ...